sdks/python/apache_beam/runners/direct/sdf_direct_runner.py

```diff
@@ -0,0 +1,264 @@
+  def new_process_fn(self, sdf):
+    return ProcessFn(
+        sdf,
+        self._process_keyed_elements_transform.ptransform_args,
+        self._process_keyed_elements_transform.ptransform_kwargs)
+
+
+class ProcessFn(beam.DoFn):
+
+  def __init__(
+      self, sdf, args_for_invoker, kwargs_for_invoker):
+    self.sdf = sdf
+    self._element_tag = _ValueStateTag('element')
+    self._restriction_tag = _ValueStateTag('restriction')
+    self.watermark_hold_tag = _ValueStateTag('watermark_hold')
+    self._process_element_invoker = None
+
+    self.sdf_invoker = DoFnInvoker.create_invoker(
+        DoFnSignature(self.sdf), context=DoFnContext('unused_context'),
+        input_args=args_for_invoker, input_kwargs=kwargs_for_invoker)
+
+  def set_step_context(self, step_context):
+    assert isinstance(step_context, DirectStepContext)
+    self._step_context = step_context
+
+  def set_process_element_invoker(self, process_element_invoker):
+    assert isinstance(process_element_invoker, SDFProcessElementInvoker)
+    self._process_element_invoker = process_element_invoker
+
+  def start_bundle(self):
+    # TODO: support start_bundle() method for SDFs.
+    pass
+
+  def finish_bundle(self):
+    # TODO: support finish_bundle() method for SDFs.
+    pass
+
+  def process(self, element, timestamp=beam.DoFn.TimestampParam,
+              window=beam.DoFn.WindowParam, * args, **kwargs):
+    if isinstance(element, KeyedWorkItem):
+      # Must be a timer firing.
+      key = element.encoded_key
+    else:
+      key, values = element
+      values = list(values)
+      assert len(values) == 1
+      value = values[0]
+
+    state = self._step_context.get_keyed_state(key)
+    element_state = state.get_state(window, self._element_tag)
+    is_seed_call = not element_state # Initially element_state is an empty list.
+
+    if not is_seed_call:
+      element = state.get_state(window, self._element_tag)
+      restriction = state.get_state(window, self._restriction_tag)
+      windowed_element = WindowedValue(element, timestamp, [window])
+    else:
+      element_and_restriction = (
+          value.value if isinstance(value, WindowedValue) else value)
+      assert isinstance(element_and_restriction, ElementAndRestriction)
+      element = element_and_restriction.element
+      restriction = element_and_restriction.restriction
+
+      if isinstance(value, WindowedValue):
+        windowed_element = WindowedValue(
+            element, value.timestamp, value.windows)
+      else:
+        windowed_element = WindowedValue(element, timestamp, [window])
+
+    tracker = self.sdf_invoker.invoke_create_tracker(restriction)
+    assert self._process_element_invoker
+    assert isinstance(self._process_element_invoker,
+                      SDFProcessElementInvoker)
+
+    output_values = self._process_element_invoker.invoke_process_element(
+        self.sdf_invoker, windowed_element, tracker)
+
+    sdf_result = None
+    for output in output_values:
```

Comment:
<!--new_thread; commit:7549b5c2ebe2ae47af9066eaf97364a27e828ab5; resolved:0-->
Can you briefly describe the protocol used for `invoke_process_element`, specifically the yielding of the last element, which is used for the residual restriction?

(see comment below regrading docstring of `invoke_process_element`.